File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/EnterRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import java.util.Optional;
import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.constant.ResultType;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.EnterRequest;
import net.prostars.messagesystem.dto.websocket.outbound.EnterResponse;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.service.ChannelService;
import net.prostars.messagesystem.service.ClientNotificationService;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class EnterRequestHandler implements BaseRequestHandler<EnterRequest> {

  private final ChannelService channelService;
  private final ClientNotificationService clientNotificationService;

  public EnterRequestHandler(
      ChannelService channelService, ClientNotificationService clientNotificationService) {
    this.channelService = channelService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, EnterRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());

    Pair<Optional<String>, ResultType> result =
        channelService.enter(request.getChannelId(), senderUserId);
    result
        .getFirst()
        .ifPresentOrElse(
            title ->
                clientNotificationService.sendMessage(
                    senderSession, senderUserId, new EnterResponse(request.getChannelId(), title)),
            () ->
                clientNotificationService.sendMessage(
                    senderSession,
                    senderUserId,
                    new ErrorResponse(MessageType.ENTER_REQUEST, result.getSecond().getMessage())));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/EnterRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.MessageType
import com.yscorp.withpush.messagesystem.constant.ResultType
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.EnterRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.EnterResponse
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.ErrorResponse
import com.yscorp.withpush.messagesystem.service.ChannelService
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession
import java.util.*

@Component
@Suppress("unused")
class EnterRequestHandler(
    private val channelService: ChannelService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<EnterRequest> {

    override fun handleRequest(senderSession: WebSocketSession, request: EnterRequest) {
        val senderUserId: UserId = senderSession.getAttributes().get(IdKey.USER_ID.value) as UserId

        val result: Pair<String?, ResultType> = channelService.enter(request.getChannelId(), senderUserId)

        val (title, errorType) = result

        if (title != null) {
            clientNotificationService.sendMessage(
                senderSession, senderUserId, EnterResponse(request.getChannelId(), title)
            )
        } else {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.ENTER_REQUEST, errorType.message)
            )
        }
    }
}
